sql: update SHOW users/roles to support provisioned users

souravcrl opened the pull request:
The SHOW users/roles view currently does not have a mechanism to filter users
via the role options that are assigned to them. This support is needed as we
will be adding `PROVISIONSRC` role option in #148200 and this needs to have
filters support. Additionally, we will be populating the last login
time(estimated) for the `system.users` table and this value will be piped here
as well. Since this value is computed on a best effort basis; it is not
guaranteed to capture every login event, and we will be adding a notice
mentioning the same.

informs #147602
fixes #147599
Epic CRDB-21590

Release note (enterprise change): The SHOW ROLES command now includes a column
that shows the estimated time that the user last logged in. Additionally,
the `options` column is now returned as an array of strings, rather than as
a single comma-separated string.

The data can be queried with a query such as:
```
root@localhost:26257/defaultdb> select * from [show roles] as r WHERE EXISTS (SELECT 1 FROM unnest(r.options) AS m(option) where option like 'SUBJECT=cn%');
    username  |            options             | member_of | estimated_last_login_time
  ------------+--------------------------------+-----------+----------------------------
    testuser  | {NOLOGIN,SUBJECT=cn=testuser}  | {admin}   | NULL
  (1 row)
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/148532)
<!-- Reviewable:end -->

souravcrl committed:
sql: update SHOW users/roles to support provisioned users

The SHOW users/roles view currently does not have a mechanism to filter users
via the role options that are assigned to them. This support is needed as we
will be adding `PROVISIONSRC` role option in #148200 and this needs to have
filters support. Additionally, we will be populating the last login
time(estimated) for the `system.users` table and this value will be piped here
as well. Since this value is computed on a best effort basis; it is not
guaranteed to capture every login event, and we will be adding a notice
mentioning the same.

informs #147602
fixes #147599
Epic CRDB-21590

Release note (enterprise change): The SHOW ROLES command now includes a column
that shows the estimated time that the user last logged in. Additionally,
the `options` column is now returned as an array of strings, rather than as
a single comma-separated string.

The data can be queried with a query such as:
```
root@localhost:26257/defaultdb> select * from [show roles] as r WHERE EXISTS (SELECT 1 FROM unnest(r.options) AS m(option) where option like 'SUBJECT=cn%');
    username  |            options             | member_of | estimated_last_login_time
  ------------+--------------------------------+-----------+----------------------------
    testuser  | {NOLOGIN,SUBJECT=cn=testuser}  | {admin}   | NULL
  (1 row)
```

souravcrl commented:
I have now added the sql notice as you had suggested @rafiss Please take a look.

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148532)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao and @pritesh-lahoti)

___
*[`pkg/sql/delegate/show_roles.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148532#-OTnLaZpEEw0a8MRnO85:-OTnLaZqDjI-fNhC47FV:bpwjaom) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f722d5c71ae56f49cc92ec8141d1ffa7d316cd/pkg/sql/delegate/show_roles.go#L18)):*
> ```Go
> func (d *delegator) delegateShowRoles() (tree.Statement, error) {
> 	sqltelemetry.IncrementShowCounter(sqltelemetry.Roles)
> 	d.evalCtx.ClientNoticeSender.BufferClientNotice(d.ctx, pgnotice.Newf(
> ```

we'll need a version gate in this function. use: `d.evalCtx.Settings.Version.IsActive(d.ctx, clusterversion.V25_3)` (assuming we will backport this change to the 25.3 branch).

if that returns false, then we should not send the notice, and we should keep using the old query from before your change.
___
*[`pkg/sql/delegate/show_roles.go` line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148532#-OTnJuM7A8x8_nLTgH_L:-OTnJuM7A8x8_nLTgH_M:b-l85xas) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f722d5c71ae56f49cc92ec8141d1ffa7d316cd/pkg/sql/delegate/show_roles.go#L19)):*
> ```Go
> 	sqltelemetry.IncrementShowCounter(sqltelemetry.Roles)
> 	d.evalCtx.ClientNoticeSender.BufferClientNotice(d.ctx, pgnotice.Newf(
> 		"estimated_last_login_time is an unreliable feature and may change in the future"))
> ```

this is a quite alarming way to phrase this sentence, since the meaning of "unreliable" is very broad. this will cause concern for users who see this message.

let's make it say: "estimated_last_login_time is computed on a best effort basis; it is not guaranteed to capture every login event"
___
*[`pkg/sql/delegate/show_roles.go` line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148532#-OTnKvyt2cUXUP4oT1Fc:-OTnKvyt2cUXUP4oT1Fd:b-samms8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f722d5c71ae56f49cc92ec8141d1ffa7d316cd/pkg/sql/delegate/show_roles.go#L25)):*
> ```Go
> 	COALESCE(array_remove(array_agg(o.option || COALESCE('=' || o.value, '') ORDER BY o.option), NULL), ARRAY[]::STRING[]) AS options,
> 	ARRAY (SELECT role FROM system.role_members AS rm WHERE rm.member = u.username ORDER BY 1) AS member_of,
> 	COALESCE(u.estimated_last_login_time::STRING, '') AS estimated_last_login_time
> ```

nit: i don't see a need to use `coalesce` for the estimated_last_login_time column. let's just allow the result to be null if it hasn't been stored yet.


<!-- Sent from Reviewable.io -->

souravcrl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148532)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao, @pritesh-lahoti, and @rafiss)

___
*[`pkg/sql/delegate/show_roles.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148532#-OTnLaZpEEw0a8MRnO85:-OTzgChfBmo97-U7wpUR:b-sxvan) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f722d5c71ae56f49cc92ec8141d1ffa7d316cd/pkg/sql/delegate/show_roles.go#L18)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

we'll need a version gate in this function. use: `d.evalCtx.Settings.Version.IsActive(d.ctx, clusterversion.V25_3)` (assuming we will backport this change to the 25.3 branch).

if that returns false, then we should not send the notice, and we should keep using the old query from before your change.
</blockquote></details>

Have address this.
___
*[`pkg/sql/delegate/show_roles.go` line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148532#-OTnJuM7A8x8_nLTgH_L:-OTzgWw06K0xDJD6XHtM:b5j8sba) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f722d5c71ae56f49cc92ec8141d1ffa7d316cd/pkg/sql/delegate/show_roles.go#L19)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this is a quite alarming way to phrase this sentence, since the meaning of "unreliable" is very broad. this will cause concern for users who see this message.

let's make it say: "estimated_last_login_time is computed on a best effort basis; it is not guaranteed to capture every login event"
</blockquote></details>

as per the discussion have updated the message.
___
*[`pkg/sql/delegate/show_roles.go` line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148532#-OTnKvyt2cUXUP4oT1Fc:-OTzgekDAlwyCxYfPdxu:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f722d5c71ae56f49cc92ec8141d1ffa7d316cd/pkg/sql/delegate/show_roles.go#L25)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: i don't see a need to use `coalesce` for the estimated_last_login_time column. let's just allow the result to be null if it hasn't been stored yet.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice! just one comment about the release note, then lgtm

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148532#-:-OUFyKe-EZUzl4SlbY_A:bjupdwo)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao and @pritesh-lahoti)

___
*[`-- commits` line 17 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148532#-OUFt9YI01cm6TPZcZ_Y:-OUFt9YI01cm6TPZcZ_Z:bibw3pv):*
nit: let's make the release note more understandable for a user.

```
Release note (enterprise change): The SHOW ROLES command now includes a column
that shows the estimated time that the user last logged in. Additionally,
the `options` column is now returned as an array of strings, rather than as
a single comma-separated string.

The data can be queried with a query such as:

 root@localhost:26257/defaultdb> select * from [show roles] as r WHERE EXISTS (SELECT 1 FROM unnest(r.options) AS m(option) where option like 'SUBJECT=cn%');
    username  |            options             | member_of | estimated_last_login_time
  ------------+--------------------------------+-----------+----------------------------
    testuser  | {NOLOGIN,SUBJECT=cn=testuser}  | {admin}   | NULL
  (1 row)
```


<!-- Sent from Reviewable.io -->

souravcrl commented:
TFTR!

bors r=rafiss

souravcrl commented:
blathers auth-check
